src/backends/r1cs/mod.rs

```diff
@@ -0,0 +1,557 @@
+            constant: None,
+        };
+
+        self.add_constraint(
+            "mul constraint: lhs * rhs = res",
+            Constraint{a, b, c},
+            span
+        );
+
+        res
+    }
+    
+    /// to constraint:
+    /// x * cst = res
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = x
+    /// b = cst
+    /// c = res
+    fn mul_const(&mut self, x: &CellVar, cst: &Self::Field, span: crate::constants::Span) -> CellVar {
+        let one = Fr::from(1);
+
+        let res = self.new_internal_var(Value::Scale(*cst, *x), span);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*x, one)])),
+            constant: None,
+        };
+
+        let b = LinearCombination {
+            terms: None,
+            constant: Some(*cst),
+        };
+
+        let c = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(res, one)])),
+            constant: None,
+        };
+
+        self.add_constraint(
+            "mul constraint: x * cst = res",
+            Constraint{a, b, c},
+            span
+        );
+
+        res
+    }
+    
+    /// to constraint:
+    /// x = cst
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = x
+    /// b = 1
+    /// c = cst
+    fn assert_eq_const(&mut self, x: &CellVar, cst: Self::Field, span: crate::constants::Span) {
+        let one = Fr::from(1);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*x, one)])),
+            constant: None,
+        };
+
+        let b = LinearCombination {
+            terms: None,
+            constant: Some(one),
+        };
+
+        let c = LinearCombination {
+            terms: None,
+            constant: Some(cst),
+        };
+
+        self.add_constraint(
+            "eq constraint: x = cst",
+            Constraint{a, b, c},
+            span
+        );
+    }
+
+    /// to constraint:
+    /// lhs = rhs
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = lhs
+    /// b = 1
+    /// c = rhs
+    fn assert_eq_var(&mut self, lhs: &CellVar, rhs: &CellVar, span: crate::constants::Span) {
+        let one = Fr::from(1);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*lhs, one)])),
+            constant: None,
+        };
+
+        let b = LinearCombination {
+            terms: None,
+            constant: Some(one),
+        };
+
+        let c = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*rhs, one)])),
+            constant: None,
+        };
+
+        self.add_constraint(
+            "eq constraint: lhs = rhs",
+            Constraint{a, b, c},
+            span
+        );
+    }
+    
+    /// Adds the public input cell vars
+    fn add_public_input(&mut self, val: Value<Self>, span: crate::constants::Span) -> CellVar {
+        let var = self.new_internal_var(val, span);
+        self.public_inputs.push(var);
```

Comment:
wondering why we track these

Reply:
It is for `SnarkjsExporter` to reorder the inputs in the witness vector.